### ISSUE TITLE
Feedback and feature requests, some small tweaks requested.

### ISSUE THREAD

**Issue by CaptainMidnight** [yde773786/Solvadoku#2]
Hi may I first congratulate you on a very promising sudoku solution assisting app. I hope you don't mind but I have some feature requests/ suggestions: -

1. Think about board orientation options i.e for tablet use where portrait only mode becomes too large and wasted space (board too big especially when keyboard masks it to enter values).
2. Cell input in current portrait only mode, can mask lowest row of cells when editing cells above them.
3. Possibly of automatically moving on to the next cell to the right when you enter a cell number.
4. Ability to save and restore a partial or board in progress i.e. enter a puzzle to solve, app accidentally closed or wrong icon pressed everything is wiped clean and you have to start again.

Having tried a lot of "sudoku solver" apps, there's a massive in the market for this app, it just needs some extra functionality to make it way better than any of the others.

Phil

**Comment by CookieCrumbs19212** [comment]
Hello Phil!

Thank you for your helpful feedback. We appreciate the time you've taken to put together a comprehensive list of items to tackle.

We're going to have a crack at Suggestions 2 to 5, the idea of having an in-app keyboard is especially interesting. 

Suggestion 1 (Creating a landscape orientation option) might take a bit longer to implement due to the fact it will require a new layout to be designed, nevertheless, we are going to start working on it ASAP.

Could you please clarify what you meant in Suggestion 6? From what I have understood: You would like an option to clear all the "solved" cells and pencil marks on the board and show only the initial Sudoku puzzle that you had entered. Please confirm if this is correct, otherwise, we would like you to further elaborate on this point.

**Comment by CaptainMidnight** [comment]
Hi @CookieCrumbs19212 

Yes your understanding of my point 6 is correct - effectively a way to revert to the as entered puzzle. Of course this option could also be looked on it like after deselecting a partial solve option it 're-loads' the original as entered puzzle.

To further explain some of my previous observations - the masking of part of the puzzle board by the onscreen keyboard overlay only appears to happen on my tablet (Samsung S5e) where the lower row of the oversized matrix can be hidden when the onscreen keyboard overlay pops up when entering numbers higher up in the matrix. So this maybe just a requirement to implement some form of layout / ratio checking as on my phone (OnePlus 6) the display is perfectly laid out, where as on the tablet (which is in forced portrait mode) the matrix takes up virtually all the screen.

I've attached a picture to help explain. Being able to use on a tablet is just a nice to have after all. In the picture I'm about to enter a number in a cell on row 8 - notice how row 9 has now been covered by the tablet google keyboard overlay ;)

[Example of google keyboard overlay covering lower row when in use on my tablet](https://www.dropbox.com/s/li47i3sl9uhqgrn/2021-08-24%2015.03.02.jpg?dl=0)

..... if you had implemented an on screen in-app keyboard this would negate the issue, there's definitely room under the sudoku matrix display for a separated row of the nine numbers....

..... also on number entry, If when entering a cell number it automatically moves to the right and at the end of a line raps to the next line or back to the top left (if you were entering a number in the bottom right), that would alllow for rapid board entry if you allow for a 10th digit i.e. a 'blank' or 'space'.

If I can help out with any testing etc just give me a shout ;)

**Comment by yde773786** [comment]
Thanks once again for the great feedback!

From a rough examination of the code I believe suggestions 3, 4 and 6 would be relatively easier to implement and would be great features in a Sudoku solver.

Locking the screen in portrait mode was a dirty trick to avoid coding up a good looking landscape layout as well. As @CookieCrumbs19212 mentioned the process would take time but we will inevitably address it in a future update.

So for suggestions 2 and 5 I think trying out a permanent keyboard is worth a shot, both for landscape and portrait mode. This would definitely avoid the covering of rows by the native keyboard, but I fear that it might clutter the bottom half of the screen. We'll come up with a couple of designs for landscape and portrait modes with a permanent keyboard and check how it comes along. We'll think of other solutions if we're not pleased with the designs but rest assured that the overlapping issue will be approached.

We will work on this project whenever we have time and I'd say first/second week of October is a realistic goal to set with regards to covering all suggestions well. 

Testing assistance would be great and we'll surely take you up on that. We'll litter this issue thread with APKs and UI designs. Tinker and comment on them as you wish. Any code contributions would be greatly appreciated as well.

Cheers!

**Comment by CaptainMidnight** [comment]
Thanks for considering them and yes completely understand time constraints etc.

Hope you don't mind, but here's a quick and dirty in-app keyboard mockup and design layout, courtesy of paint ;)

[App design layout mockup including in-app keyboard](https://www.dropbox.com/s/psuqx4opm0edikm/keyboard%20mockup.jpg?dl=0)

**Comment by CookieCrumbs19212** [comment]
So here is the current draft we have drawn up for the new Solvadoku UI. Its not the final UI design, we are still taking inputs and changing features.

@CaptainMidnight Please let us know what you think of it and what changes you would make.
![Solvadoku UI - Portrait 2](https://user-images.githubusercontent.com/62071086/131248015-d30cc059-f397-4b24-bd55-720101942e95.png)
![Solvadoku UI - Portrait 3](https://user-images.githubusercontent.com/62071086/131248020-57ea175f-5137-4db5-8855-466e10df2b36.png)

**Comment by CaptainMidnight** [comment]
> So here is the current draft we have drawn up for the new Solvadoku UI. Its not the final UI design, we are still taking inputs and changing features.
> 
> @CaptainMidnight Please let us know what you think of it and what changes you would make.


It's great to see both a portrait and landscape design, just for that alone well done. If my understanding is correct I'm assuming the green right arrow button would input a space in a cell when entering an empty cell .... or is it to move onto another cell - the arrow symbol I don't think makes it immediately clear compared to a '_' symbol. That's down to interpretation I suppose as possibly every time you press a number button you may have/will have implemented the automatic move to the next cell functionality - greatly increasing sudoku starting matrix setup time, so to input an empty cell would be the green arrow?

If I was being super critical I would look at the spacing of the buttons in the landscape version. I like the slight format change of the 3x option buttons, but that section could move slightly further up, but still be below the sudoku matrix edge and allow for slightly bigger and even spaced 12x round buttons i.e the 12x buttons on portrait layout spacing should be copied to the landscape layout. If you look closely the clearer/bigger 12x button layout in my opinion is smaller and more cramped on the landscape layout - like the function button change though, that's cool :)

For a precise view look at the positioning of the 12x button section size and position compared to its location relative to the sudoku matrix and the 3x function button section - in landscape it's been reduced, length and width making it look more squashed.

Usability for people with dexterity issues could potentially be reduced in landscape as a side effect....

Please don't get me wrong these work in progress improvements are thoroughly appreciated.  ;)

**Comment by CookieCrumbs19212** [comment]
Yup, I can definitely make the Number Pad bigger, there's plenty of whitespace around the buttons, and as you mentioned, it would improve the usability in landscape mode. 

For reference, below is the terminology we use for the elements in our layout, this will help us better communicate and understand each other in the future:

![UI Terminology Legend](https://user-images.githubusercontent.com/62071086/131526717-d798a197-9005-479b-9965-bd8aa1b048af.png)

I will rework the landscape draft with the following points:
1. Move the 3 Action button further up
2. Resize the Numberpad with bigger keys.

**Comment by CaptainMidnight** [comment]
Hi @CookieCrumbs19212 is there any side loadable .apk available to see what this is currently looking like on my tablet (testing out the orientation rotation) or are we still some way off with an interim 1st look beta version?

**Comment by CookieCrumbs19212** [comment]
Hey Phil!

We have not completed landscape mode yet, but we have already started work on it. We expect it to be ready within the week (if there aren't too many bugs slowing us down). We just have to perfect the button layout and do some code refactoring and it should be good to go. We took your input and made the keypad buttons larger and also pushed the action buttons a bit higher.

From the initial list of suggestions you gave us, we have completed points 2, 3, and 5. We also made some other UI updates and bug fixes. 

Unfortunately, we don't think we will be able to implement points 4 and 6 in this update but we have decided to gradually roll out these features with minor updates in the very near future. We expect
![Solvadoku UI - Landscape 6](https://user-images.githubusercontent.com/62071086/135827416-072375bf-c95e-4a48-b259-138de196c275.png)
 to release our next update within the next 2 weeks.

As for an APK, I will certainly share one once we've finished the landscape mode, we'd be interested in your feedback because we do not have a physical tablet to test out landscape mode, rather, we've solely depended on the android emulator for testing.

Here is the mock-up on which we are basing the landscape mode.

**Comment by CookieCrumbs19212** [comment]
@CaptainMidnight,
Hello Phil

We have added landscape mode to the app. We will be completing some last-minute brush-ups and roll out the update this weekend, which is today or tomorrow for you depending on your timezone, and you will be able to update the app directly from the Play Store within a few days. 

I am not sure how uploading an APK to GitHub would work but if you would still like one, let me know, I'll figure something out.
